### PR TITLE
PC-1354: Validate entered postcodes on address search

### PR DESCRIPTION
**Pull request by Glenn-Clarke**
[Link to Jira ticket](https://beisdigital.atlassian.net/browse/PC-1354)

# Description

Added HUG2 postcode validation to address entry. 
This prevents partial postcodes from being entered, narrowing the property search.

# Checklist

- [x] I have made any necessary updates to the documentation
- [x] I have checked there are no unnecessary IDE warnings in this PR
- [x] I have checked there are no unintentional line ending changes
- [x] I have added tests where applicable
- [x] If I have made any changes to the code, I have used the IDE auto-formatter on it
- [x] If I have made any changes to Python files (even if not changing any content strings), I have run `make extract-translations`
- [x] If I have made any changes to website flow, I have updated the [Flow Miro Board](https://miro.com/app/board/uXjVK4AgQ3E=/)
- [x] If I have made any changes to website flow, I have checked forward and back behaviour is still consistent

# Screenshots

{F952A3CD-9FBB-4E80-B353-80E83108F753}

{01FB09BA-ADCA-46D0-9DD2-20896935B557}

### CODE REVIEW

File: help_to_heat/frontdoor/schemas.py [hunk 1]
```diff
@@ -853,6 +857,20 @@ class SessionSchema(Schema):
...
+            value = value.upper().replace(" ", "")
+            valid_postcode = False
+
+            for valid_pattern in postcode_regex_collection:
+                if re.match(valid_pattern, value):
+                    valid_postcode = True
+                    break
+
+            if not valid_postcode:
+                raise ValidationError([_("Enter a valid UK postcode")])
```
Comment (jdgage) [inline help_to_heat/frontdoor/schemas.py:872]: Could condense this and make it a little clearer by inverting the check:
```suggestion
            if not any(map(lambda pattern: re.match(pattern, value), postcode_regex_collection)):
                raise ValidationError([_("Enter a valid UK postcode")])
```

**Review by jdgage — CHANGES_REQUESTED** [review]

File: help_to_heat/frontdoor/schemas.py (hunk 1 above)
Comment (jdgage) [reply help_to_heat/frontdoor/schemas.py:872]: Yes was going to say there was probably a more Python way to write the same thing

File: help_to_heat/frontdoor/schemas.py [hunk 2]
```diff
@@ -853,6 +857,20 @@ class SessionSchema(Schema):
     epc_found = fields.String()
     epc_rating_is_eligible = fields.String()
 
+    @validates("postcode")
+    def validate_postcode(self, value):
+        if value:
+            value = value.upper().replace(" ", "")
+            valid_postcode = False
+
+            for valid_pattern in postcode_regex_collection:
```
[the remark this answers is not in the document]
Comment (jdgage) [reply help_to_heat/frontdoor/schemas.py:866]: Agree + would drop "collection" as that's self evident

File: help_to_heat/frontdoor/schemas.py [hunk 3]
```diff
@@ -853,6 +857,20 @@ class SessionSchema(Schema):
     epc_found = fields.String()
     epc_rating_is_eligible = fields.String()
 
+    @validates("postcode")
+    def validate_postcode(self, value):
+        if value:
+            value = value.upper().replace(" ", "")
```
[the remark this answers is not in the document]
Comment (Glenn-Clarke) [reply help_to_heat/frontdoor/schemas.py:863]: It doesn't appear to, no. 
Is there a situation where a user could get that character in there? (I guess if they copy/paste?)

[the remark this answers is not in the document]
Comment (Glenn-Clarke) [reply help_to_heat/frontdoor/schemas.py]: I've changed how the replacement is done, so it picks up other whitespace characters like tabs. I've also moved the upper after the replacement for a (negligible) efficiency saving.

File: help_to_heat/frontdoor/schemas.py [hunk 4]
```diff
@@ -860,15 +860,8 @@ class SessionSchema(Schema):
...
-            value = value.upper().replace(" ", "")
-            valid_postcode = False
-
-            for valid_pattern in postcode_regex_collection:
-                if re.match(valid_pattern, value):
-                    valid_postcode = True
-                    break
-
-            if not valid_postcode:
+            standardised_value = re.sub(r'\s', '', value).upper()
```
Comment (jdgage) [inline help_to_heat/frontdoor/schemas.py:863]: 👍

**Review by jdgage — APPROVED** [review]

**Pull request merged**